util/econftool.c

```diff
@@ -0,0 +1,557 @@
+char path_filename[PATH_MAX]; /* the path concatenated with the filename and the suffix */
+char root_dir[PATH_MAX] = "/etc";
+char usr_root_dir[PATH_MAX] = "/usr/etc";
+char xdg_config_dir[PATH_MAX];
+
+/**
+ * @brief Shows the usage.
+ */
+
+static void usage(void)
+{
+    fprintf(stderr, "Usage: %s COMMANDS [OPTIONS] filename.conf\n\n", utilname);
+    fprintf(stderr, "COMMAND:\n");
+    fprintf(stderr, "show     reads all snippets for filename.conf and prints all groups,\n");
+    fprintf(stderr, "         keys and their values.\n");
+    fprintf(stderr, "edit     starts the editor $EDITOR (environment variable) where the\n");
+    fprintf(stderr, "         groups, keys and values can be modified and saved afterwards.\n");
+    fprintf(stderr, "  -f, --full:      copy the original configuration file to /etc instead of\n");
+    fprintf(stderr, "                   creating drop-in files.\n");
+    fprintf(stderr, "  -y, --yes:       Assumes yes for all prompts.\n");
+    fprintf(stderr, "revert   reverts all changes to the vendor versions. Basically deletes\n");
+    fprintf(stderr, "         the config file and snippet directory in /etc.\n");
+    fprintf(stderr, "  -y, --yes:       Assumes yes for all prompts and runs non-interactively.\n\n");
+}
+
+/**
+ * @brief Change root dir if enviroment variable "ECONFTOOL_ROOT" exists
+ *
+ * @param path The path to be changed
+ */
+static void change_root_dir(char path[PATH_MAX])
+{
+    if (getenv("ECONFTOOL_ROOT") != NULL) {
+        if (strlen(path) + strlen(getenv("ECONFTOOL_ROOT")) >= PATH_MAX) {
+            fprintf(stderr, "ECONFTOOL_ROOT + path is too long\n");
+            exit(EXIT_FAILURE);
+        }
+
+        char *tmp = strdup(path);
+
+        strcpy(path, getenv("ECONFTOOL_ROOT"));
+        strcat(path, tmp);
+
+        free(tmp);
+    }
+}
+
+/**
+ * @brief This function is used as a callback for nftw to delete
+ *        all files nftw finds.
+ */
+static int nftw_remove(const char *path, const struct stat *sb, int flag, struct FTW *ftwbuf)
+{
+    return remove(path);
+}
+
+/**
+ * @brief Generate a tmpfile using mkstemp() and write its name into tmp_name
+ */
+static int generate_tmp_file(char *tmp_name)
+{
+    int filedes;
+    char name_template[21] = "/tmp/econftool-XXXXXX";
+    filedes = mkstemp(name_template);
+    if (filedes == -1) {
+        perror("mkstemp() failed");
+        return -1;
+    }
+
+    char fdpath[PATH_MAX];
+    char tmp_filename[PATH_MAX];
+    snprintf(fdpath, PATH_MAX, "/proc/self/fd/%d", filedes);
+    ssize_t r = readlink(fdpath, tmp_filename, PATH_MAX);
+    tmp_filename[r] = '\0';
+    strcpy(tmp_name, tmp_filename);
+    return 0;
+}
+
+/**
+ * @brief This command will read all snippets for filename.conf
+ *        (econf_readDirs) and print all groups, keys and their
+ *        values as an application would see them.
+ */
+static int econf_show(void)
+{
+    econf_err econf_error;
```

Comment:
This function would benefit from goto-style error handling or the \_\_cleanup__ C extension